File: data_structures/binary_trees/breadth_first_traversal.js
// Implementation of Breadth-First Search Algorithm

class Node {
    constructor(val) {
        this.val = val;
        this.left = null;
        this.right = null;
    }
}

const breadthFirstValues = (root) => {

    if (root === null) return [];
    const queue = [ root ];
    const result = [];

    while (queue.length > 0) {

        const current = queue.shift();
        result.push(current.val);
 
        if (current.left !== null) queue.push(current.left);
        if (current.right !== null) queue.push(current.right);

    }
    return result;

}


const a = new Node("a");
const b = new Node("b");
const c = new Node("c");
const d = new Node("d");
const e = new Node("e");
const f = new Node("f");

a.left = b;
a.right = c;
b.left = d;
b.right = e;
c.right = f;

console.log(breadthFirstValues(a));

File: data_structures/binary_trees/breadth_first_traversal.py
# Implementation of Breadth-First Search Algorithm

from collections import deque

class Node(): 

    def __init__(self, val):
        self.val = val
        self.left = None
        self.right = None



def breadth_first_values(root):

    if root is None:
        return []
    
    result = []
    queue = deque([root])

    while queue:
        current = queue.popleft()
        result.append(current.val)

        if current.left is not None:
            queue.append(current.left)
        if current.right is not None:
            queue.append(current.right)


    return result


a = Node("a")
b = Node("b")
c = Node("c")
d = Node("d")
e = Node("e")
f = Node("f")

a.left = b
a.right = c
b.left = d
b.right = e
c.right = f

print(breadth_first_values(a))